### PR TITLE
CORDA-3155 contractStateTypeMappings in the NodeVaultService is not thread safe

### PR DESCRIPTION
**Pull request by josecoll**
https://r3-cev.atlassian.net/browse/CORDA-3155

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 1]
```diff
@@ -748,7 +754,7 @@ class NodeVaultService(
     }
 
     private fun <T : ContractState> deriveContractTypes(clazz: Class<T>): Set<Class<T>> {
-        val myTypes : MutableSet<Class<T>> = mutableSetOf()
+        val myTypes : MutableSet<Class<T>> = CopyOnWriteArraySet()
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: I don't think this is the right change.  This set gets mapped in the two places it gets called.  Was fine as it was (I think, but maybe I missed something) but in each of the two places it gets mapped, a call is made to `getOrPut` on `contractStateTypeMappings` with a `mutableSetOf`, and both those call sites should switch that out for `CopyOnWriteArraySet`

Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Lines 109 and 746 in this file.

**Review by rick-r3 — CHANGES_REQUESTED** [review]

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt (hunk 1 above)
Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: Hmmm .... the exception is a `ConcurrentModificationException` on an iterator.

I have attempted to line up the code line number given by the stack trace:
`net.corda.node.services.vault.NodeVaultService$start$1.call(NodeVaultService.kt:808) ~[corda-node-4.0.20190
603_openj9.jar:?]`

but these do not line up  (the above referenced jar appears to be a modified version of patch 4.0.20190603).

The nearest iterator to that line number is given by:

```
        clazz.interfaces.forEach {
            if (it != ContractState::class.java) {
                myTypes.add(uncheckedCast(it))
                myTypes.addAll(deriveContractTypes(uncheckedCast(it)))
            }
        }

```

within `deriveContractTypes`.

Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: I am awaiting for the log files from the customer - who has confirmed via Slack that the patch (as per this code) is now working.

Comment (josecoll) [reply node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: I'll apply your suggestion as well (belt and braces approach).

**Comment by EnterpriseBlockChainSolutions** [comment]
Ok.bllkk the
Ivy ty ❤️❤️ nuh jn.

On Aug 28, 2019 2:24 PM, josecoll <notifications@github.com> wrote:

@josecoll commented on this pull request.

________________________________

In node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt<https://github.com/corda/corda/pull/5389#discussion_r318577446>:

> @@ -748,7 +754,7 @@ class NodeVaultService(
     }

     private fun <T : ContractState> deriveContractTypes(clazz: Class<T>): Set<Class<T>> {
-        val myTypes : MutableSet<Class<T>> = mutableSetOf()
+        val myTypes : MutableSet<Class<T>> = CopyOnWriteArraySet()


I am awaiting for the log files from the customer - who has confirmed via Slack that the patch (as per this code) is now working.

**Comment by josecoll** [comment]
Re-targeting specified branch version as per JIRA ticket.
`master` is now frozen following branching strategy changes.

**Pull request closed without merging**